File: pycli/crmscli.py
# Copyright 2018 Wenzhe Liu (liuwenzhe2008@qq.com)
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import etcd3
import json
import thread
import os
import threading
from etcd3.events import PutEvent
from etcd3.events import DeleteEvent

JOB_PREFIX = 'crms/job/'
SERVER_PREFIX = 'crms/server/'
ASSIGN_PREFIX = 'crms/assign/'


def synchronized(func):
    func.__lock__ = threading.Lock()

    def synced_func(*args, **kws):
        with func.__lock__:
            return func(*args, **kws)

    return synced_func


class JobState(object):
    def __init__(self):
        self.status = 'new'
        self.stdouterr = ''


class Job(object):
    def __init__(self):
        self.id = ''
        self.command = []  # jobCommand is an array of each part of the command
        self.state = JobState()

    def get_state(self):
        return self.state


def start_server(server_host, name, parellel_count, etcd_host_port):
    ''' no wait for started '''
    os.system('ssh %s GoCRMS %s %d %s &' %(server_host, name, parellel_count, etcd_host_port))


def start_server_by_lsf(name, parellel_count, etcd_host_port):
    ''' no wait for started '''
    os.system(
        # 'bsub -q pwodebug "GoCRMS %s %d %s &"'
        'bsub -R "type==any" -q pwodebug "GoCRMS %s %d %s" &'
        % (name, parellel_count, etcd_host_port))


def start_multiple_servers_by_lsf(server_count, name_prefix, parellel_count, etcd_host_port):
    ''' no wait for started '''
    os.system(
        # 'bsub -q pwodebug "GoCRMS %s %d %s &"'
        'bsub -R "type==any" -q pwodebug "gocrmsstarter %d %s %d %s" &'
        % (server_count, name_prefix, parellel_count, etcd_host_port))


class CrmsCli(object):
    def __init__(self, host_port='localhost:2379'):
        hp = host_port.split(':')
        host = hp[0]
        port = int(hp[1])
        self.cli = etcd3.client(host, port)
        self.__servers = {}  # key: server name, value: server parellel job count
        self.__cancel_watch_servers = None
        self.__jobs = {}  # key: jobId, value: Job
        self.__cancelWatchJobs = None
        self.__onJobStatusChanged = None

        self.get_servers()
        self.get_jobs()

    def add_watcher(self, on_job_status_changed):
        self.get_servers()
        self.get_jobs()

        self.__onJobStatusChanged = on_job_status_changed

    def close(self):
        if self.__cancel_watch_servers is not None:
            self.__cancel_watch_servers()
            self.__cancel_watch_servers = None
        if self.__cancelWatchJobs is not None:
            self.__cancelWatchJobs()
            self.__cancelWatchJobs = None
        print "close"

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    @synchronized
    def get_servers(self):
        if self.__cancel_watch_servers is None:  # not watch yet
            self.__servers = self.__get_servers()
            self.__watch_servers()
        return self.__servers

    def __get_servers(self):
        servers = self.cli.get_prefix(SERVER_PREFIX)
        return {wk[1].key[len(SERVER_PREFIX):]: wk[0] for wk in servers}

    def __watch_servers(self):
        evts, self.__cancel_watch_servers = self.cli.watch_prefix(SERVER_PREFIX)

        def update_servers(evts):
            for evt in evts:
                if isinstance(evt, PutEvent):
                    self.__servers[evt.key[len(SERVER_PREFIX):]] = evt.value
                elif isinstance(evt, DeleteEvent):
                    self.__servers.pop(evt.key[len(SERVER_PREFIX):])

        thread.start_new_thread(update_servers, (evts,))

    def stop_server(self, name):
        self.cli.put(SERVER_PREFIX + name, 'close')

    def stop_servers(self):
        for server in self.get_servers().keys():
            self.stop_server(server)

    def create_job(self, job_id, job_command):  # jobCommand is an array of each part of the command
        cmd = json.dumps(job_command)  # e.g: ['ls', '-l', '..']
        self.cli.put(JOB_PREFIX + job_id, cmd)

    def run_job(self, job_id, server):
        self.cli.put(ASSIGN_PREFIX + server + '/' + job_id, '')

    def __get_job_or_create_if_absent(self, job_id):
        if not self.__jobs.has_key(job_id):
            job = Job()
            job.id = job_id
            self.__jobs[job_id] = job
        return self.__jobs[job_id]

    @synchronized
    def __update_job(self, k, v):
        ks = k.split('/')[2:]
        n = len(ks)
        job_id = ks[0]
        job = self.__get_job_or_create_if_absent(job_id)
        if n == 1:
            job.id = job_id
            job.command = json.loads(v)
        elif n == 2:
            prop = ks[1]
            if prop == 'state':
                job.state.status = v
                if self.__onJobStatusChanged is not None:
                    self.__onJobStatusChanged(job)
        elif n == 3:
            prop = ks[2]
            if prop == "stdouterr":
                job.state.stdouterr = v

    def __get_jobs(self):
        ''' example of key-value format in etcd server:
        job/3
        ["ls", "-l", ".."]
        job/3/state
        done
        job/3/state/stdouterr
        total 1760
        drwxr-xr-x 1 weliu 1049089       0 Dec 13 17:18 angular
        drwxr-xr-x 1 weliu 1049089       0 Jan 17 16:53 bctools
        drwxr-xr-x 1 weliu 1049089       0 Jan  2 09:47 cluster
        '''
        jobs = self.cli.get_prefix(JOB_PREFIX)
        for job in jobs:
            k = job[1].key
            v = job[0]
            self.__update_job(k, v)
        return self.__jobs

    def __watch_jobs(self):
        events, self.__cancelWatchJobs = self.cli.watch_prefix(JOB_PREFIX)

        def update_jobs(evts):
            for evt in evts:
                if isinstance(evt, PutEvent):
                    self.__update_job(evt.key, evt.value)
                elif isinstance(evt, DeleteEvent):
                    pass  # currently no job remove yet

        thread.start_new_thread(update_jobs, (events,))

    @synchronized
    def get_jobs(self):
        if self.__cancelWatchJobs is None:
            self.__get_jobs()
            self.__watch_jobs()
        return self.__jobs

    @synchronized
    def get_job(self, job_id):
        jobs = self.get_jobs()
        return jobs[job_id]

    def nodes(self):
        result = self.cli.get_prefix('crms/')
        nodes = {}
        for r in result:
            k = r[1].key
            v = r[0]
            nodes[k] = v
        return nodes

    def clean(self):
        ''' clean assign and job, not clean server '''
        self.cli.delete_prefix('crms/assign/')
        self.cli.delete_prefix('crms/job/')


# example of usage
if __name__ == "__main__":
    cli = CrmsCli()
    # import pdb
    # pdb.set_trace()
    print cli.get_servers()
    # import time

    # time.sleep(20)
    # print cli.getservers()
    cli.create_job("1234", ['pwd'])
    cli.run_job("1234", "wenzhe")


    def print_jobs(jobs):
        for jobId, job in jobs.items():
            print "job id:", job.id
            print "job command", ' '.join(job.command)
            print "job status", job.state.status
            print job.state.stdouterr

    print_jobs(cli.get_jobs())
    cli.run_job("1234", "qiqi")
    print_jobs(cli.get_jobs())
    cli.close()

File: gocrmscli/crmscli.go
// Copyright 2018 Wenzhe Liu (liuwenzhe2008@qq.com)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package gocrmscli

import (
	"context"
	"encoding/json"
	"log"
	"strconv"
	"strings"
	"time"

	"github.com/coreos/etcd/clientv3"
	"errors"
)

const (
	crmsPrefix   = "crms/"
	serverPrefix = crmsPrefix + "server/"
	jobPrefix    = crmsPrefix + "job/"
	assignPrefix = crmsPrefix + "assign/"

	WATCH_EVT_PUT    = 0
	WATCH_EVT_DELETE = 1

	STATUS_NEW     = "new"
	STATUS_RUNNING = "running"
	STATUS_DONE    = "done"
	STATUS_FAIL    = "fail"
)

type JobState struct {
	Status    string
	Stdouterr string
}

type Job struct {
	ID      string
	Command []string             // Job Command is an array of each part of the command
	State   JobState // key: assigned server name, value: JobData (state + stdout/err)
}

func NewJob(id string) *Job {
	return &Job{
		ID:      id,
		Command: make([]string, 0),
		State:   JobState{STATUS_NEW, ""},
	}
}

func (job *Job) GetStatus() string {
	return job.State.Status
}

func (job *Job) GetStdOutErr() string {
	return job.State.Stdouterr
}

type Server struct {
	Name            string
	ParellelAbility int
	// the job current assign to the server, server is now handling it or waiting for handling.
	// it is used as a set, the value is useless (as Go doesn't have set)
	Jobs map[string]bool
}

func NewServer(name string, parellelAbility int) *Server {
	return &Server{name, parellelAbility, make(map[string]bool)}
}

type CrmsCli struct {
	cli                *clientv3.Client
	requestTimeout     time.Duration
	servers            map[string]*Server // key: server name, value: server parellel job count
	jobs               map[string]*Job    // key: jobId, value: Job
	cancelWatchServers context.CancelFunc
	cancelWatchJobs    context.CancelFunc
	cancelWatchAssign  context.CancelFunc
}

func New(endpoints []string, dialTimeout time.Duration, requestTimeout time.Duration) (crms *CrmsCli, err error) {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   endpoints,
		DialTimeout: dialTimeout,
	})
	if err != nil {
		return
	}
	crms = &CrmsCli{
		cli:            cli,
		requestTimeout: requestTimeout,
		servers:        make(map[string]*Server),
		jobs:           make(map[string]*Job),
	}
	return
}

func (crms *CrmsCli) Close() {
	if crms.cancelWatchServers != nil {
		crms.cancelWatchServers()
	}
	if crms.cancelWatchJobs != nil {
		crms.cancelWatchJobs()
	}
	if crms.cancelWatchAssign != nil {
		crms.cancelWatchAssign()
	}
	crms.cli.Close()
}

func (crms *CrmsCli) GetServer(name string) (server *Server, exist bool, err error) {
	servers, err := crms.GetServers()
	if err != nil {
		return
	}
	server, exist = servers[name]
	return
}

func (crms *CrmsCli) GetServers() (servers map[string]*Server, err error) {
	if crms.cancelWatchServers == nil {
		crms.servers, err = crms.getServers()
		if err != nil {
			return
		}
		crms.watchServers() //TODO: if the server change before watch event start, how to do?
	}
	if crms.cancelWatchAssign == nil {
		crms.watchAssign()
	}
	servers = crms.servers
	return
}

func (crms *CrmsCli) getServers() (servers map[string]*Server, err error) {
	resp, err := crms.get(serverPrefix, clientv3.WithPrefix())
	if err != nil {
		log.Println(err)
		return
	}
	servers = make(map[string]*Server)
	for _, kv := range resp.Kvs {
		serverName := string(kv.Key)[len(serverPrefix):]
		v := string(kv.Value)
		if v == "close" {
			continue
		}
		parellelAbility, err := strconv.Atoi(v)
		if err != nil {
			log.Println(err)
			continue
		}
		server := NewServer(serverName, parellelAbility)
		jobIds, err := crms.GetAssignJobs(serverName)
		if err != nil {
			log.Println(err)
			continue
		}
		server.Jobs = make(map[string]bool)
		for _, jobId := range jobIds {
			server.Jobs[jobId] = true
		}
		servers[serverName] = server
	}
	return
}

func (crms *CrmsCli) watchServers() {
	rch, cancel := crms.watch(serverPrefix, clientv3.WithPrefix())
	crms.cancelWatchServers = cancel
	go func(servers map[string]*Server) {
		for wresp := range rch {
			for _, ev := range wresp.Events {
				name := string(ev.Kv.Key)[len(serverPrefix):]
				switch ev.Type {
				case WATCH_EVT_PUT:
					v := string(ev.Kv.Value)
					if v == "close" {
						log.Println("delete server", name)
						delete(servers, name)
						continue
					}
					parellelAbility, err := strconv.Atoi(v)
					if err != nil {
						log.Println(err)
						continue
					}
					server, exist := servers[name]
					if exist {
						server.ParellelAbility = parellelAbility
					} else {
						server := NewServer(name, parellelAbility)
						servers[name] = server
					}
				case WATCH_EVT_DELETE:
					log.Println("delete server", name)
					delete(servers, name)
				default:
					log.Printf("Unknown event type %s (%q : %q)\n", ev.Type, ev.Kv.Key, ev.Kv.Value)
				}
			}
		}
	}(crms.servers)
}

func (crms *CrmsCli) StopServer(name string) error {
	return crms.put(serverPrefix+name, "close")
}

// jobCommand is an array of each part of the command
func (crms *CrmsCli) CreateJob(jobId string, jobCommand []string) error {
	cmd, err := json.Marshal(jobCommand)
	if err != nil {
		return err
	}
	crms.put(jobPrefix+jobId, string(cmd))
	return nil
}

func (crms *CrmsCli) RunJob(jobId string, server string) error {
	return crms.put(assignPrefix+server+"/"+jobId, "")
}

func (crms *CrmsCli) getJobOrCreateIfAbsent(jobId string) *Job {
	job, exist := crms.jobs[jobId]
	if !exist {
		job = NewJob(jobId)
		crms.jobs[jobId] = job
	}
	return job
}

// k starts with "crms/job/<jobid>"
func (crms *CrmsCli) updateJob(k string, v []byte) error {
	ks := strings.Split(k, "/")[2:]
	n := len(ks)
	jobId := ks[0]
	job := crms.getJobOrCreateIfAbsent(jobId)
	switch n {
	case 1:
		job.ID = jobId
		if err := json.Unmarshal(v, &job.Command); err != nil {
			return err
		}
	case 2:
		if ks[1] == "state" {
			job.State.Status = string(v)
		}
	case 3:
		if ks[1] == "state" && ks[2] == "stdouterr" {
			job.State.Stdouterr = string(v)
		}
	}
	return nil
}

// example of key-value format in etcd server:
//    job/3
//    ["ls", "-l", ".."]
//    job/3/state
//    done
//    job/3/state/stdouterr
//    total 1760
//    drwxr-xr-x 1 weliu 1049089       0 Dec 13 17:18 angular
//    drwxr-xr-x 1 weliu 1049089       0 Jan 17 16:53 bctools
//    drwxr-xr-x 1 weliu 1049089       0 Jan  2 09:47 cluster
func (crms *CrmsCli) getJobs() (map[string]*Job, error) {
	resp, err := crms.get(jobPrefix, clientv3.WithPrefix())
	if err != nil {
		return crms.jobs, err
	}
	for _, kv := range resp.Kvs {
		k := string(kv.Key)
		v := kv.Value
		crms.updateJob(k, v)
	}
	return crms.jobs, nil
}

func (crms *CrmsCli) watchJobs() {
	rch, cancel := crms.watch(jobPrefix, clientv3.WithPrefix())
	crms.cancelWatchJobs = cancel
	go func() {
		for wresp := range rch {
			for _, ev := range wresp.Events {
				switch ev.Type {
				case WATCH_EVT_PUT:
					err := crms.updateJob(string(ev.Kv.Key), ev.Kv.Value)
					if err != nil {
						log.Println(err)
					}
				case WATCH_EVT_DELETE:
					// currently no job remove yet
				default:
					log.Printf("Unknown event type %s (%q : %q)\n", ev.Type, ev.Kv.Key, ev.Kv.Value)
				}
			}
		}
	}()
}

func (crms *CrmsCli) GetJobs() (jobs map[string]*Job, err error) {
	if crms.cancelWatchJobs == nil {
		jobs, err = crms.getJobs()
		if err != nil {
			return
		}
		crms.watchJobs()
	} else {
		jobs = crms.jobs
	}
	return
}

func (crms *CrmsCli) GetJob(jobId string) (job *Job, exist bool, err error) {
	jobs, err := crms.getJobs()
	if err != nil {
		return
	}
	job, exist = jobs[jobId]
	return
}

// return the assign jobs' id that currently assign to the server (not including finish)
func (crms *CrmsCli) GetAssignJobs(server string) (jobs []string, err error) {
	prefix := assignPrefix + server + "/"
	n := len(prefix)
	resp, err := crms.get(prefix, clientv3.WithPrefix())
	if err != nil {
		return
	}
	jobs = make([]string, len(resp.Kvs))
	for i, kv := range resp.Kvs {
		jobId := string(kv.Key)[n:]
		jobs[i] = jobId
	}
	return
}

func (crms *CrmsCli) GetJobsByServer(server string) (jobs []*Job, err error) {
	assignJobIds, err := crms.GetAssignJobs(server)
	if err != nil {
		return
	}
	if len(assignJobIds) == 0 {
		return
	}
	js, err := crms.GetJobs()
	if err != nil {
		return
	}
	jobs = make([]*Job, 0, len(assignJobIds))
	for i, jobId := range assignJobIds {
		job, exist := js[jobId]
		if !exist {
			err = errors.New("job id " + jobId + " doesn't exist")
			return
		}
		jobs[i] = job
	}
	return
}

func (crms *CrmsCli) watchAssign() {
	rch, cancel := crms.watch(assignPrefix, clientv3.WithPrefix())
	crms.cancelWatchAssign = cancel
	go func(servers map[string]*Server) {
		for wresp := range rch {
			for _, ev := range wresp.Events {
				k := string(ev.Kv.Key) // k starts with crms/assign/<server_name>
				ks := strings.Split(k, "/")
				if len(ks) < 3 { // invalid
					log.Println("Invalid key", k)
					continue
				}
				ks = ks[2:]
				serverName := ks[0]
				server, exist := crms.servers[serverName] //TODO: may not thread safe
				if !exist {
					// nothing need to update
					continue
				}
				jobId := ks[1]
				if ev.Type == WATCH_EVT_DELETE {
					delete(server.Jobs, jobId)
				} else if ev.Type == WATCH_EVT_PUT {
					server.Jobs[jobId] = true
				}
			}
		}
	}(crms.servers)
}

// TODO: the following common method is duplicate with server.go.
func (crms *CrmsCli) put(key, val string, opts ...clientv3.OpOption) error {
	ctx, cancel := context.WithTimeout(context.Background(), crms.requestTimeout)
	defer cancel()
	_, err := crms.cli.Put(ctx, key, val, opts...)
	if err != nil {
		log.Println(err)
	}
	return err
}

func (crms *CrmsCli) get(key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), crms.requestTimeout)
	defer cancel()
	return crms.cli.Get(ctx, key, opts...)
}

func (crms *CrmsCli) watch(key string, opts ...clientv3.OpOption) (clientv3.WatchChan, context.CancelFunc) {
	ctx, cancel := context.WithCancel(context.Background())
	return crms.cli.Watch(ctx, key, opts...), cancel
}

func (crms *CrmsCli) delete(key string, opts ...clientv3.OpOption) (*clientv3.DeleteResponse, error) {
	ctx, cancel := context.WithTimeout(context.Background(), crms.requestTimeout)
	defer cancel()
	return crms.cli.Delete(ctx, key, opts...)
}

func (crms *CrmsCli) GetNodes() (nodes map[string]string, err error) {
	resp, err := crms.get(crmsPrefix, clientv3.WithPrefix())
	if err != nil {
		return
	}
	nodes = make(map[string]string)
	for _, kv := range resp.Kvs {
		k := string(kv.Key)
		v := string(kv.Value)
		nodes[k] = v
	}
	return
}
